recipes/openai.py

```diff
@@ -0,0 +1,391 @@
+    openai_n_retries: int
+    openai_n: int
+    examples: List[PromptExample]
+    response_parser: Callable
+
+    OPENAI_COMPLETIONS_ENDPOINT: str = "https://api.openai.com/v1/completions"
+    RETRY_ERROR_CODES: List[str] = [429, 503]
+
+    def __init__(
+        self,
+        prompt_template: jinja2.Template,
+        *,
+        labels: List[str],
+        max_examples: int,
+        segment: bool,
+        openai_api_org: str,
+        openai_api_key: str,
+        openai_model: str,
+        response_parser: Callable,
+        openai_temperature: int = 0,
+        openai_max_tokens: int = 500,
+        openai_retry_timeout_s: int = 1,
+        openai_read_timeout_s: int = 30,
+        openai_n_retries: int = 10,
+        openai_n: int = 1,
+        verbose: bool = False,
+    ):
+        self.prompt_template = prompt_template
+        self.model = openai_model
+        self.labels = [normalize_label(label) for label in labels]
+        self.max_examples = max_examples
+        self.verbose = verbose
+        self.segment = segment
+        self.examples = []
+        self.openai_api_org = openai_api_org
+        self.openai_api_key = openai_api_key
+        self.openai_temperature = openai_temperature
+        self.openai_max_tokens = openai_max_tokens
+        self.openai_retry_timeout_s = openai_retry_timeout_s
+        self.openai_read_timeout_s = openai_read_timeout_s
+        self.openai_n = openai_n
+        self.openai_n_retries = openai_n_retries
+        self.response_parser = response_parser
+
+    def __call__(
+        self,
+        stream: Iterable[Dict],
+        *,
+        nlp: Language,
+        batch_size: int,
+        **kwargs,
+    ) -> Iterable[Dict]:
+        if self.segment:
+            stream = preprocess.split_sentences(nlp, stream)
+
+        stream = self.pipe(stream, nlp, batch_size, **kwargs)
+        stream = self.set_hashes(stream)
+        return stream
+
+    def pipe(self, stream: Iterable[Dict], nlp: Language, batch_size: int, **kwargs):
+        """Process the stream and add suggestions from OpenAI."""
+        stream = self.stream_suggestions(stream, batch_size)
+        stream = self.format_suggestions(stream, nlp=nlp)
+        return stream
+
+    def set_hashes(self, stream: Iterable[Dict]):
+        for example in stream:
+            yield set_hashes(example)
+
+    def update(self, examples: Iterable[Dict]) -> float:
+        """Update the examples that will be used in the prompt based on user flags."""
+        for eg in examples:
+            if PromptExample.is_flagged(eg):
+                self.add_example(PromptExample.from_prodigy(eg, self.label))
+        return 0.0
+
+    def add_example(self, example: PromptExample) -> None:
+        """Add an example for use in the prompts. Examples are pruned to the most recent max_examples."""
+        if self.max_examples:
+            self.examples.append(example)
+        if len(self.examples) >= self.max_examples:
```

Comment:
Change `>=` to `>`. If it's equal, no need to prune right? Or even remove the len based condition and always `self.examples = self.examples[-self.max_examples :]`. Not sure which is more efficient.

Reply:
Updated to use `>`